File: calendar.R
library(httr)
library(jsonlite)
library(gargle)
library(dplyr)
library(tibble)
library(ellmer)

USER_EMAIL <- "tool.call.demo@gmail.com"

calendar_authenticate <- function() {
  my_app <- gargle_oauth_client_from_json("credentials-demo.json")

  token_fetch(
    scopes = "https://www.googleapis.com/auth/calendar.readonly",
    app = my_app,
    email = USER_EMAIL,
    cache = ".secrets"
  )
}

query_api <- function(start_date, end_date, token) {
  res <-
    GET(
      url = "https://www.googleapis.com/calendar/v3/calendars/primary/events",
      config(token = token),
      query = list(
        timeMin = glue::glue("{start_date}T00:00:00Z"),
        timeMax = glue::glue("{end_date}T23:59:59Z"),
        singleEvents = "true",
        orderBy = "startTime"
      )
    )

  content(res, as = "parsed", simplifyVector = TRUE)$items
}

#' Gets the calendar events between start_date and end_date
#'
#' @param start_date Start date, in YYYY-MM-DD format.
#' @param end_date End date, in YYYY-MM-DD format.
#' @return A tibble containing the calendar events in the date range. 
get_calendar_events <- function(start_date, end_date) {

  token <- calendar_authenticate()

  events <- query_api(start_date, end_date, token)

  if (length(events) == 0) { return(tibble()) }

  tibble(
    id = events$id,
    summary = events$summary,
    start = events$start$dateTime,
    end =  events$end$dateTime
  )
}

#' Gets the current date
#'
#' @return The current date, as a Date object. 
get_date <- function() {
  Sys.Date()
}

chat <- chat_claude(system_prompt = readLines("prompt-calendar.md"))

chat$register_tool(tool(
  get_calendar_events,
  "Fetches calendar events between a specified start and end date.",
  start_date = type_string(
       "The start date from which to fetch calendar events. It should be a string representing the 
date in the format YYYY-MM-DD.",
       required = TRUE
  ),
  end_date = type_string(
       "The end date up to which to fetch calendar events. It should be a string representing the 
date in the format YYYY-MM-DD.",
       required = TRUE
  )
))

chat$register_tool(tool(get_date, "Gets the user's current date."))


live_console(chat)

File: calendar.py
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
import chatlas
from typing import List, Dict
from datetime import date
import os

SCOPES = ["https://www.googleapis.com/auth/calendar.readonly"]
CREDENTIALS_FILE = "credentials-demo.json"
TOKEN_FILE = "token.json"

def get_date() -> str:
    """
    Get the current date.

    Returns
    -------
    str
        The current date in "YYYY-MM-DD" format.
    """
    return date.today().isoformat()

def authenticate():
    creds = None

    if os.path.exists(TOKEN_FILE):
        creds = Credentials.from_authorized_user_file(TOKEN_FILE, SCOPES)

    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                CREDENTIALS_FILE,
                scopes=SCOPES
            )
            creds = flow.run_local_server(port=0)
            with open(TOKEN_FILE, "w") as token:
                token.write(creds.to_json())

    return creds

# def authenticate():
#     flow = InstalledAppFlow.from_client_secrets_file(CREDENTIALS_FILE, SCOPES)
#     return flow.run_local_server(port=0)

def build_calendar_service(creds):
    return build("calendar", "v3", credentials=creds)

def query_api(service, start_date, end_date):
    time_min = f"{start_date}T00:00:00Z"
    time_max = f"{end_date}T23:59:59Z"

    events = service.events().list(
        calendarId="primary",
        timeMin=time_min,
        timeMax=time_max,
        singleEvents=True,
        orderBy="startTime"
    ).execute().get("items", [])

    return [
        {
            "start": e["start"].get("dateTime", e["start"].get("date")),
            "summary": e.get("summary", "No title")
        }
        for e in events
    ]

def get_calendar_events(start_date: str, end_date: str) -> List[Dict[str, str]]:
    """
    Fetch Google Calendar events between two dates.

    Parameters
    ----------
    start_date : str
        The start date in "YYYY-MM-DD" format.
    end_date : str
        The end date in "YYYY-MM-DD" format.

    Returns
    -------
    List[dict]
        A list of events with 'start' and 'summary' fields.
    """
    creds = authenticate()
    service = build_calendar_service(creds)
    return query_api(service, start_date, end_date)


chat = chatlas.ChatAnthropic(system_prompt=open("prompt-calendar.md", "r").read())

chat.register_tool(get_calendar_events)
chat.register_tool(get_date)
chat.console()

